sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java

```diff
@@ -541,23 +542,37 @@ public <T> Output<T> invoke(String token, TypeShape<T> targetType, InvokeArgs ar
                 ? CompletableFuture.completedFuture(null)
                 : packageRef;
 
-            // Find all the resource dependencies from dependsOn. We need to wait for these futures to complete
-            // before calling the invoke.
-            var depsFuture = this.prepare.getAllTransitivelyReferencedResourceUrnsAsync(ImmutableSet.copyOf(options.getDependsOn()));
+            // The expanded set of dependencies, including children of components.
+            var transitiveDeps = this.prepare.getAllTransitivelyReferencedResources(ImmutableSet.copyOf(options.getDependsOn()));
+            // If we depend on any CustomResources, we need to ensure that their
+            // ID is known before proceeding. If it is not known, we will return
+            // an unknown result.
+            var hasUnknownIDs = CompletableFutures.allOf(transitiveDeps
+                .filter(r -> r instanceof CustomResource)
+                .map(r -> (CustomResource) r)
+                .map(r -> Internal.of(r.id()).isKnown())
```

Comment:
Does Java give you stream fusion for free? If not then we might want to combine these `map`s into one that just does `r -> Internal.of((CustomResource) r).id()).isKnown()` or something like that.